### PR TITLE
listener: envoy.reloadable_features.listener_wildcard_match_ip_family deprecation

### PR DESCRIPTION
**Pull request by daixiang0**
Signed-off-by: Loong <loong.dai@intel.com>

Commit Message: listener: envoy.reloadable_features.listener_wildcard_match_ip_family deprecation
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] #19507
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/server/connection_handler_impl.cc [hunk 1]
```diff
@@ -252,27 +252,17 @@ ConnectionHandlerImpl::getBalancedHandlerByAddress(const Network::Address::Insta
...
+      address.ip()->version() == Network::Address::IpVersion::v4
+          ? Network::Utility::getIpv4AnyAddress(address.ip()->port())->asString()
+          : Network::Utility::getIpv6AnyAddress(address.ip()->port())->asString();
+
+  auto iter = tcp_listener_map_by_address_.find(addr_str);
+  if (iter != tcp_listener_map_by_address_.end()) {
+    details = *iter->second;
   }
+
   return (details.has_value())
```
Comment (soulxu) [inline source/server/connection_handler_impl.cc:266]: Actually, you can get rid of the temporary variable `details` and the line 262 to 264.

Then you can write return it something as below:
```
return iter != tcp_listener_map_by_address_.end()? Network::BalancedConnectionHandlerOptRef(
                   ActiveTcpListenerOptRef(absl::get<std::reference_wrapper<ActiveTcpListener>>(
                                               *iter->second->typed_listener_))
                       .value()
                       .get())
           : absl::nullopt
```

**Review by soulxu — CHANGES_REQUESTED** [review]

File: source/server/connection_handler_impl.cc (hunk 1 above)
Comment (daixiang0) [reply source/server/connection_handler_impl.cc:266]: This var makes code logic clearer than removing it.

**Comment by jacob-delgado** [comment]
This should not be removed until https://github.com/envoyproxy/envoy/issues/18134 is resolved.

**Comment by daixiang0** [comment]
still available.

**Comment by soulxu** [comment]
since https://github.com/envoyproxy/envoy/pull/19716 is already merged, this pr can be merged now.

you probably need to merge with the main branch.

**Comment by daixiang0** [comment]
I will do that.

**Comment by ggreenway** [comment]
Looks like you missed a change in removing the runtime flag, which is causing it to not compile.

/wait

**Comment by daixiang0** [comment]
I will fix it.

File: source/server/connection_handler_impl.cc [hunk 2]
```diff
@@ -313,23 +313,24 @@ ConnectionHandlerImpl::getBalancedHandlerByAddress(const Network::Address::Insta
...
   // We do not return stopped listeners.
   // TODO(wattli): consolidate with previous search for more efficiency.
-  std::string addr_str =
-      address.ip()->version() == Network::Address::IpVersion::v4
-          ? Network::Utility::getIpv4AnyAddress(address.ip()->port())->asString()
-          : Network::Utility::getIpv6AnyAddress(address.ip()->port())->asString();
+
+  std::string addr_str = address.ip()->version() == Network::Address::IpVersion::v4
+                             ? Network::Address::Ipv4Instance(address.ip()->port()).asString()
+                             : Network::Address::Ipv6Instance(address.ip()->port()).asString();
```
Comment (soulxu) [inline source/server/connection_handler_impl.cc:319]: Why not use `Utility::getIpv4AnyAddress` any more? It seems more efficient than build an new any address Instance.

**Comment by soulxu** [comment]
lgtm.

**Review by ggreenway — APPROVED** [review]

**Pull request merged**